### ISSUE TITLE
Default JDK not updated after changing java.home

### ISSUE THREAD

**Issue by hkuadithya** [redhat-developer/vscode-java#1288]
Issue Type: <b>Bug</b>

VSCode Java extension seems to raise error while using `var` data type (type inference for local variables was introduced way back with Java 10). Running the project also seems to fail while using `var` datatype. 

There's absolutely no problem with my Java environment. Compiling and running the same Java file using `javac` and `java` succeeds in command line. This works fine with IntelliJ IDEA as well. **Please check the screenshots of the error and Configure java runtime.**

**Is there some setting or configuration I need to change?**

Java environment details:
```
java 13.0.2 2020-01-14
Java(TM) SE Runtime Environment (build 13.0.2+8)
Java HotSpot(TM) 64-Bit Server VM (build 13.0.2+8, mixed mode, sharing)
```
### Error
![image](https://user-images.githubusercontent.com/20955088/74703160-25076100-51c1-11ea-9b43-b737d3958e40.png)

### Configure java runtime - JDK 13
![image](https://user-images.githubusercontent.com/20955088/74703126-0903bf80-51c1-11ea-8187-d7814d09ec2b.png)

---

Extension version: 0.8.1
VS Code version: Code 1.42.1 (c47d83b293181d9be64f27ff093689e8e7aed054, 2020-02-11T14:45:59.656Z)
OS version: Windows_NT x64 10.0.18362

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz (4 x 2904)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: enabled<br>rasterization: enabled<br>skia_renderer: disabled_off<br>surface_control: disabled_off<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.91GB (8.79GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>

**Comment by fbricon** [comment]
are you using a Maven or Gradle project? or simple Java files in a folder?

**Comment by hkuadithya** [comment]
@fbricon It's just a simple standalone Java file with a main method. Not using Gradle/Maven.

**Comment by fbricon** [comment]
did you simply open the java file or its parent folder?

**Comment by hkuadithya** [comment]
`File > Open Folder` (empty folder)
Created a new Java file from the Explorer in the left panel.  `New File > Functional.java`.
`Debug > Run Without debugging` -- Works fine. However, when I add `var`, errors are raised.

Examples for errors:
```
var array = new int[]{1, 2, 3, 4};
var hello = "hello world";
```

This works fine:
```
int[] array = {1, 2 ,3, 4};
String hello = "hello world";
```

**Comment by fbricon** [comment]
Please turn on [verbose logging](https://github.com/redhat-developer/vscode-java/wiki/Troubleshooting#turn-on-message-tracing-between-vs-code-and-the-java-language-server) and attach the logs from the output view, after restarting vscode

**Comment by fbricon** [comment]
It works for me on Mac. @snjeza can you double check on windows?

**Comment by hkuadithya** [comment]
@fbricon Sure, I'm currently at work. Will provide more details (verbose logging) when I get back home in about 6-7 hours (PST timezone).

Since it works on Mac, I suspect it's specific to the OS. (Win 10 Professional, 64 bit).

**Comment by snjeza** [comment]
> @snjeza can you double check on windows?

I can't reproduce the issue on Windows and Linux.
@hkuadithya Could you attach your Functional.java?

**Comment by hkuadithya** [comment]
#### Functional.java
```
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class Functional {
    public static void main(String[] args) {
        var a = "hello";
        var b = new int[]{1, 2, 3, 4, 5};
        testMap();
        testFilter();
        testReduce();
    }

    private static void testMap() {
        List<String> list = Arrays.asList("abc", "def", "ghi");
        List<String> result = list.stream().map((element) -> element.toUpperCase()).collect(Collectors.toList());
        System.out.println(result);
    }

    private static void testFilter() {
        int[] elements = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
        int[] x = Arrays.stream(elements).filter((element) -> ((element & 1) == 0)).toArray();

        System.out.println(Arrays.toString(x));
        System.out.println(Arrays.toString(x));
    }

    private static void testReduce() {
        int[] elements = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
        int product = Arrays.stream(elements).reduce(0, (left, right) -> left + right);
        System.out.println(product);
    }
}
```

![image](https://user-images.githubusercontent.com/20955088/74959562-a7f60a80-53bf-11ea-9d49-376f5818232f.png)

#### Verbose Logging trace
```
{
  message: "[Trace - 9:09:06 AM] Sending request 'workspace/executeCommand - (3)'.",
  level: 'info',
  timestamp: '2020-02-20 09:09:06.383'
}
{
  message: 'Params: {\n    "command": "vscode.java.updateDebugSettings",\n    ' +
    '"arguments": [\n        ' +
    '"{\\"showHex\\":false,\\"showStaticVariables\\":false,\\"showQualifiedNames\\":false,\\"showLogicalStructure\\":true,\\"showToString\\":true,\\"maxStringLength\\":0,\\"numericPrecision\\":0,\\"hotCodeReplace\\":\\"manual\\",\\"enableRunDebugCodeLens\\":true,\\"forceBuildBeforeLaunch\\":true,\\"console\\":\\"integratedTerminal\\",\\"logLevel\\":\\"WARNING\\",\\"javaHome\\":\\"C:\\\\\\\\Program ' +
    'Files\\\\\\\\Java\\\\\\\\jdk-13.0.2\\"}"\n    ]\n}\n\n',
  level: 'info',
  timestamp: '2020-02-20 09:09:06.384'
}
{
  message: '[Trace - 9:09:06 AM] Received response ' +
    "'workspace/executeCommand - (3)' in " +
    '30ms.',
  level: 'info',
  timestamp: '2020-02-20 09:09:06.413'
}
{
  message: 'Result: ' +
    '"{\\"maxStringLength\\":0,\\"numericPrecision\\":0,\\"showStaticVariables\\":false,\\"showQualifiedNames\\":false,\\"showHex\\":false,\\"showLogicalStructure\\":true,\\"showToString\\":true,\\"logLevel\\":\\"WARNING\\",\\"javaHome\\":\\"C:\\\\\\\\Program ' +
    'Files\\\\\\\\Java\\\\\\\\jdk-13.0.2\\",\\"hotCodeReplace\\":\\"manual\\"}"\n' +
    '\n',
  level: 'info',
  timestamp: '2020-02-20 09:09:06.413'
}
{
  message: "[Trace - 9:09:06 AM] Sending request 'java/buildWorkspace - (4)'.",
  level: 'info',
  timestamp: '2020-02-20 09:09:06.415'
}
{
  message: "[Trace - 9:09:06 AM] Received notification 'window/logMessage'.",
  level: 'info',
  timestamp: '2020-02-20 09:09:06.428'
}
{
  message: 'Params: {\n    "type": 1,\n    "message": "Feb 20, 2020, 9:09:06 AM ' +
    'Error occured while building workspace. Details: \\r\\n message: ' +
    'var cannot be resolved to a type; code: 16777218; resource: ' +
    'C:/Users/hkuad/Desktop/CP/Practice/Functional.java; line: 7\\r\\n ' +
    'message: var cannot be resolved to a type; code: 16777218; ' +
    'resource: C:/Users/hkuad/Desktop/CP/Practice/Functional.java; ' +
    'line: 8"\n}\n\n',
  level: 'info',
  timestamp: '2020-02-20 09:09:06.428'
}
{
  message: '[Error - 9:09:06 AM] Feb 20, 2020, 9:09:06 AM Error occured while ' +
    'building workspace. Details: \r' +
    '\n message: var cannot be resolved to a type; code: 16777218; resource: ' +
    'C:/Users/hkuad/Desktop/CP/Practice/Functional.java; line: 7\r' +
    '\n message: var cannot be resolved to a type; code: 16777218; resource: ' +
    'C:/Users/hkuad/Desktop/CP/Practice/Functional.java; line: 8',
  level: 'info',
  timestamp: '2020-02-20 09:09:06.429'
}
{
  message: '[Trace - 9:09:06 AM] Received response ' +
    "'java/buildWorkspace - (4)' in 15ms.",
  level: 'info',
  timestamp: '2020-02-20 09:09:06.430'
}
{
  message: 'Result: 2\n\n',
  level: 'info',
  timestamp: '2020-02-20 09:09:06.430'
}
{
  message: "[Trace - 9:09:06 AM] Received notification 'language/progressReport'.",
  level: 'info',
  timestamp: '2020-02-20 09:09:06.527'
}
{
  message: 'Params: {\n    "id": ' +
    '"b0db6b5d-9f4e-464b-8c63-a68522abbf59",\n    "task": ' +
    '"Building",\n    "status": "0% ",\n    "totalWork": 1000,\n   ' +
    ' "workDone": 0,\n    "complete": false\n}\n\n',
  level: 'info',
  timestamp: '2020-02-20 09:09:06.528'
}
{
  message: "[Trace - 9:09:06 AM] Received notification 'language/progressReport'.",
  level: 'info',
  timestamp: '2020-02-20 09:09:06.531'
}
{
  message: 'Params: {\n    "id": ' +
    '"b0db6b5d-9f4e-464b-8c63-a68522abbf59",\n    "task": ' +
    '"Building",\n    "status": "0% ",\n    "totalWork": 1000,\n  ' +
    '  "workDone": 0,\n    "complete": true\n}\n\n',
  level: 'info',
  timestamp: '2020-02-20 09:09:06.532'
}
```

**Comment by fbricon** [comment]
Can you paste the full log, starting when vscode-java starts?

**Comment by hkuadithya** [comment]
@fbricon 

This is the only trace I got when VSCode and Language server start.
```
{
  message: 'Starting Java server with: C:\\Program Files\\Java\\jdk-13.0.2\\bin\\java ' +
    '--add-modules=ALL-SYSTEM --add-opens java.base/java.util=ALL-UNNAMED ' +
    '--add-opens java.base/java.lang=ALL-UNNAMED ' +
    '-Declipse.application=org.eclipse.jdt.ls.core.id1 ' +
    '-Dosgi.bundles.defaultStartLevel=4 ' +
    '-Declipse.product=org.eclipse.jdt.ls.core.product ' +
    '-Dfile.encoding=utf8 -DwatchParentProcess=false -Xmx1G -XX:+UseG1GC ' +
    '-XX:+UseStringDeduplication -jar ' +
    'c:\\Users\\hkuad\\.vscode\\extensions\\redhat.java-0.57.0\\server\\plugins\\org.eclipse.equinox.launcher_1.5.700.v20200207-2156.jar ' +
    '-configuration ' +
    'c:\\Users\\hkuad\\.vscode\\extensions\\redhat.java-0.57.0\\server\\config_win ' +
    '-data ' +
    'c:\\Users\\hkuad\\AppData\\Roaming\\Code\\User\\workspaceStorage\\c96b088e4efde1a7a0114c8383b5b363\\redhat.java\\jdt_ws',
  level: 'info',
  timestamp: '2020-02-20 09:15:34.266'
}
```

**Comment by fbricon** [comment]
Can you attach the content of c:\Users\hkuad\AppData\Roaming\Code\User\workspaceStorage\c96b088e4efde1a7a0114c8383b5b363\redhat.java\jdt_ws\.metadata\.plugins\org.eclipse.jdt.launching (a couple xml files)

**Comment by snjeza** [comment]
@hkuadithya You can also try to call Java: Clean the Java language server workspace

**Comment by hkuadithya** [comment]
@fbricon @snjeza 
Yes, `Java: Clean the Java language server workspace` worked.
I suppose it's some language server caching issue. I also recently updated my JDK from JDK9 to JDK13. Not sure if that caused this problem. 

Let me know if you guys need anything else for the RCA of this issue.

**Comment by fbricon** [comment]
@snjeza the default JDK should have been updated to JDK 13, so something's still not right here. Please investigate.

**Comment by fbricon** [comment]
To make sure I'm not jumping the guns, @hkuadithya how did you change the JDK version exactly? via java.home?

**Comment by hkuadithya** [comment]
@fbricon 
Maybe this will help you reproduce the issue:-
1. Before updating JDK: I created a new folder and added new file `Functional.java`

2. I was able to build and run without any issues. Since it was JDK9, my code didn't have any `var`data type

3. I uninstalled old Java, updated my JDK to 13, updated the path in environment variables and restarted the system

4. Opened VSCode, and added `var` to code to test type inference. Build and compilation failed. `javac` worked though

5. I thought that vscode wasn't picking the right Java version or maybe JDK13 wasn't supported yet since it was released recently.

6. I checked "Configure Java runtime"and saw that JDK13 path was already present under option 4 (source: other). 

7. I added the same path for option 1 under user settings json hoping that it would force the language server to pick the features from latest JDK. Restarted vscode and tested. That didn't work either.

TLDR; vscode had automatically detected the jdk path under "other". I also added the path to option 1 "Java.home" hoping that language server would refresh.

**Comment by snjeza** [comment]
I can't reproduce the issue.
See https://www.screencast.com/t/Wivjt8ccVES

**Comment by hkuadithya** [comment]
@fbricon @snjeza 
Looks like this is an isolated case/issue which occurs under some weird set of circumstances.
I couldn't find any other issues resembling this under [Open Issues](https://github.com/redhat-developer/vscode-java/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+var). You folks could focus more time and resources on this if similar issues are opened in future.

I will close this issue for now. Thank you for investigating this. 
Please reopen if this warrants further investigation.
